src/backend/CassandraBackendNew.h

```diff
@@ -0,0 +1,943 @@
+        {
+            auto const seq = forward ? rng->minSequence : rng->maxSequence;
+            auto const placeHolder =
+                forward ? 0u : std::numeric_limits<std::uint32_t>::max();
+
+            statement.bindAt(1, std::make_tuple(placeHolder, placeHolder));
+            log_.debug() << "account = " << ripple::strHex(account)
+                         << " idx = " << seq << " tuple = " << placeHolder;
+        }
+
+        // FIXME: Limit is a hack to support uint32_t properly for the time
+        // being. Should be removed later and schema updated to use proper
+        // types.
+        statement.bindAt(2, Limit{limit});
+        auto const res = executor_.read(yield, statement);
+        auto const& results = res.value();
+        if (not results.hasRows())
+        {
+            log_.debug() << "No rows returned";
+            return {};
+        }
+
+        std::vector<ripple::uint256> hashes = {};
+        auto numRows = results.numRows();
+        log_.info() << "num_rows = " << numRows;
+
+        for (auto [hash, data] :
+             extract<ripple::uint256, std::tuple<uint32_t, uint32_t>>(results))
+        {
+            hashes.push_back(hash);
+            if (--numRows == 0)
+            {
+                log_.debug() << "Setting cursor";
+                cursor = data;
+
+                // forward queries by ledger/tx sequence `>=`
+                // so we have to advance the index by one
+                if (forward)
+                    ++cursor->transactionIndex;
+            }
+        }
+
+        auto const txns = fetchTransactions(hashes, yield);
+        log_.debug() << "Txns = " << txns.size();
+
+        if (txns.size() == limit)
+        {
+            log_.debug() << "Returning cursor";
+            return {txns, cursor};
+        }
+
+        return {txns, {}};
+    }
+
+    bool
+    doFinishWrites() override
+    {
+        // wait for other threads to finish their writes
+        executor_.sync();
+
+        if (!range)
+        {
+            executor_.writeSync(
+                schema_->updateLedgerRange,
+                ledgerSequence_,
+                false,
+                ledgerSequence_);
+        }
+
+        if (not executeSyncUpdate(schema_->updateLedgerRange.bind(
+                ledgerSequence_, true, ledgerSequence_ - 1)))
+        {
+            log_.warn() << "Update failed for ledger " << ledgerSequence_;
+            return false;
+        }
+
+        log_.info() << "Committed ledger " << ledgerSequence_;
+        return true;
+    }
+
+    void
+    writeLedger(ripple::LedgerInfo const& ledgerInfo, std::string&& header)
+        override
+    {
+        executor_.write(
+            schema_->insertLedgerHeader, ledgerInfo.seq, std::move(header));
+
+        executor_.write(
+            schema_->insertLedgerHash, ledgerInfo.hash, ledgerInfo.seq);
+
+        // FIXME: should this be atomic or protected by lock??
+        ledgerSequence_ = ledgerInfo.seq;
```

Comment:
pretty sure this is only accessed by a single thread at a time. IIRC it's mainly used so you don't need to pass the ledger sequence into `doFinishWrites`. But probably worth double checking

Reply:
Thanks for letting me know. Since `tsan` is not screaming at us anymore, i think this is called from one thread indeed. I will double check tho.